pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+from pubnub.models.consumer.common import PNStatus
+from pubnub.workers import BaseMessageWorker
+
+
+class Effect:
+    pubnub: PubNub = None
+    event_engine = None
+    invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]
+    stop_event = None
+    logger: logging.Logger
+    task: asyncio.Task
+
+    def set_pn(self, pubnub: PubNub):
+        self.pubnub = pubnub
+
+    def __init__(self, pubnub_instance, event_engine_instance,
+                 invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]) -> None:
+        self.invocation = invocation
+        self.event_engine = event_engine_instance
+        self.pubnub = pubnub_instance
+
+        self.logger = logging.getLogger("pubnub")
+
+    def run(self):
+        pass
+
+    def run_async(self, coro):
+        loop: asyncio.AbstractEventLoop = self.pubnub.event_loop
+        if loop.is_running():
+            self.task = loop.create_task(coro)
+        else:
+            self.task = loop.run_until_complete(coro)
+
+    def stop(self):
+        if self.stop_event:
+            self.logger.debug(f'stop_event({id(self.stop_event)}).set() called on {self.__class__.__name__}')
+            self.stop_event.set()
+        if hasattr(self, 'task') and isinstance(self.task, asyncio.Task) and not self.task.cancelled():
+            self.task.cancel()
+
+    def get_new_stop_event(self):
+        event = asyncio.Event()
+        self.logger.debug(f'creating new stop_event({id(event)}) for {self.__class__.__name__}')
+        return event
+
+    def calculate_reconnection_delay(self, attempts):
+        if self.reconnection_policy is PNReconnectionPolicy.LINEAR:
+            delay = self.interval
+
+        elif self.reconnection_policy is PNReconnectionPolicy.EXPONENTIAL:
+            delay = int(math.pow(2, attempts - 5 * math.floor((attempts - 1) / 5)) - 1)
+        return delay
+
+
+class HandshakeEffect(Effect):
+    def run(self):
+        channels = self.invocation.channels
+        groups = self.invocation.groups
+        tt = self.invocation.timetoken or 0
+        if hasattr(self.pubnub, 'event_loop'):
+            self.stop_event = self.get_new_stop_event()
+            self.run_async(self.handshake_async(channels=channels,
+                                                groups=groups,
+                                                timetoken=tt,
+                                                stop_event=self.stop_event))
+
+    async def handshake_async(self, channels, groups, stop_event, timetoken: int = 0):
+        request = Subscribe(self.pubnub).channels(channels).channel_groups(groups).cancellation_event(stop_event)
+
+        if feature_enabled('PN_MAINTAIN_PRESENCE_STATE') and hasattr(self.pubnub, 'state_container'):
+            state_container = self.pubnub.state_container
+            request.state(state_container.get_state(channels))
+
+        request.timetoken(0)
```

Comment:
Do you need to set the `timetoken` to zero in each of the occurrences? It would be nice to have a single point of truth in case of code-reviewing-and-debugging. 
In C# there was so many of the reassignments like that and it caused the bug. :/